### PR TITLE
Fail when trying to shove object dtypes into bcolz

### LINKED ISSUE #181: Bcolz conversion with a dshape containing object dtype should fail
```python
In [1]: ct = odo('odo/backends/tests/iris.csv', 'iris.bcolz')

In [2]: ct
ValueError: chunkfile iris.bcolz/species/data/__-149.blp not found
```

We should not allow object dtypes on conversion.

### PR DESCRIPTION
**Pull request by cpcloud**
closes #181

### CODE REVIEW

**Comment by cpcloud** [comment]
moving to 0.4.0 as this is a backwards incompatible change: bcolz doesn't support object dtypes well

**Comment by cpcloud** [comment]
@mrocklin does this break any of your existing code that might be appending object dtyped arrays to a bcolz store?

**Comment by mrocklin** [comment]
I've learned to avoid object dtype arrays and bcolz.  No issues from me.

File: odo/backends/bcolz.py [hunk 1]
```diff
@@ -82,6 +101,23 @@ def resource_bcolz(uri, dshape=None, expected_dshape=None, **kwargs):
...
+                                "If you have strings in those columns, pass in"
+                                " a fixed length string dshape for each of "
+                                "those columns.\nFor example, if the maximum "
+                                "length of each string in column 'a' is 10"
+                                "instead of var * {a: int32, b: string}, "
+                                "pass var * {a: int32, b: string[10, 'A']}" %
+                                [name for (name, _), keep in
+                                 zip(dtype_pairs, is_object) if keep])
+        else:
+            if dt.type == np.object_:
```
Comment (llllllllll) [inline odo/backends/bcolz.py]: `elif`

File: odo/backends/bcolz.py [hunk 2]
```diff
@@ -82,6 +101,22 @@ def resource_bcolz(uri, dshape=None, expected_dshape=None, **kwargs):
         dshape = datashape.dshape(dshape)
 
         dt = dshape_to_numpy(dshape)
+        if isrecord(dshape.measure):
+            dtype_pairs = [(name, dt[name].type) for name in dt.names]
+            is_object = [dtype == np.object_ for _, dtype in dtype_pairs]
+            if any(is_object):
+                raise TypeError("object dtypes are not supported by bcolz.\n"
```
Comment (FrancescAlted) [inline odo/backends/bcolz.py]: Well, object dtypes are actually supported, but I agree that they are actually not very usable unless the objects are large, in which case they are quite fine. Maybe you can change that message by "small object dtypes are poorly supported by bcolz"?

Comment (cpcloud) [inline odo/backends/bcolz.py]: Here's what I'm talking about:

```
In [18]: from bcolz import ctable

In [19]: import pandas as pd

In [20]: recs = pd.DataFrame(dict(a=list('abcd'), b=np.random.randint(10, size=4))).to_records(index=False).view(np.ndarray)

In [21]: recs
Out[21]:
array([('a', 2), ('b', 7), ('c', 2), ('d', 6)],
      dtype=[('a', 'O'), ('b', '<i8')])

In [22]: c = ctable(np.array([], dtype=recs.dtype), rootdir='df')

In [23]: c.append(recs)

In [24]: c[:2]
ValueError: could not broadcast input array from shape (4) into shape (2)

In [25]: c
ValueError: chunkfile df/a/data/__-3.blp not found
```

vs no object dtypes, which works fine:

```
In [1]: from bcolz import ctable

In [2]: import pandas as pd

In [3]: recs = pd.DataFrame(dict(a=np.arange(4), b=np.random.randint(10, size=4))).to_records(index=False).view(np.ndarray)

In [4]: c = ctable(np.array([], dtype=recs.dtype), rootdir='df')

In [5]: c.append(recs)

In [6]: c
Out[6]:
ctable((4,), [('a', '<i8'), ('b', '<i8')])
  nbytes: 64; cbytes: 32.00 KB; ratio: 0.00
  cparams := cparams(clevel=5, shuffle=True, cname='blosclz')
  rootdir := 'df'
[(0, 3) (1, 4) (2, 4) (3, 6)]

In [7]: c[:2]
Out[7]:
array([(0, 3), (1, 4)],
      dtype=[('a', '<i8'), ('b', '<i8')])
```

Comment (cpcloud) [inline odo/backends/bcolz.py]: Looks like the fix is on the way for this:

https://github.com/Blosc/bcolz/pull/242

Comment (cpcloud) [inline odo/backends/bcolz.py]: I'd be thrilled to fully can this PR and let object dtypes through

Comment (FrancescAlted) [inline odo/backends/bcolz.py]: Ah, I see.  Then proceed as you prefer.

**Comment by cpcloud** [comment]
closing, as this was an actual bug in bcolz. we'll continue to allow appending object dtypes for the foreseeable future

**Pull request closed without merging**